### PR TITLE
Refactor controls

### PR DESCRIPTION
**Pull request by ayambabu23**
Every time we want to add a new key, we just need to add it to the key list in `gui.py`. The `KeySignal` class will handle the rest.

### CODE REVIEW

File: controls_pyqt/gui.py [hunk 1]
```diff
@@ -212,24 +213,25 @@ def setup_ui(self):
     def start_ui(self):
         sys.exit(self.app.exec_())
 
-    def on_trigger(self, trigger: str):
+    def on_trigger(self, key: str, pressed: bool):
+        trigger = KeySignal(key, pressed)
         self.comms.add_to_queue(trigger)
```
Comment (priime0) [inline controls_pyqt/gui.py:218]: Is `comms` refactored to handle `KeySignal`s?

**Review by priime0 — APPROVED** [review]

**Review by lutetjeff — CHANGES_REQUESTED** [review]
The code is fine and ready to merge, but please rework it a bit for the updated main.
- Update control binds in GUI (maybe look at diff vs main?)
- Please remake your changes on the new version of `comms.py`

File: comms.py [hunk 2]
```diff
@@ -83,30 +83,32 @@ def read_send(self):
                 continue
 
             # get multiplier
-            multiplier_percent = self.initial_percent if len(key_pressed) == 1 else \
-                                              (0 if key_pressed[0] == "s" else self.MULTIPLIER_PERCENT)
+            key_is_released = not key_pressed.pressed
+            multiplier_percent = self.initial_percent if(key_pressed.pressed and not shift) else \
```
Comment (lutetjeff) [inline comms.py]: `key_pressed.shift`

File: controls_pyqt/key_signal.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Class that stores each key press as an object
+
+class KeySignal:
+	def __init__(self, key: str, pressed: bool):
+		self.key: str = key
+		self.pressed: bool = pressed
+		
+		if self.key ^ 32 > self.key or self.key ^ 32 == 0: # self.key ^ 32 == 0 checks whether it is a space or not
```
Comment (lutetjeff) [inline controls_pyqt/key_signal.py]: in what world do you XOR a string and a number?
what does this even do?

Comment (lutetjeff) [reply controls_pyqt/key_signal.py]: maybe compare the `ord(self.key)` value instead?

File: controls_pyqt/gui.py [hunk 4]
```diff
@@ -212,24 +213,25 @@ def setup_ui(self):
...
 
         if not key_event.isAutoRepeat():
-            for (key, trigger) in self.KEYS_PRESSED:
+            for key in self.KEYS_PRESSED:
                 # print(key, trigger)
                 if key_event.text() == key:
-                    self.on_trigger(trigger)
+                    self.on_trigger(key, True)
 
     def keyReleaseEvent(self, keyevent):
```
Comment (lutetjeff) [inline controls_pyqt/gui.py:229]: typo

**Comment by lutetjeff** [comment]
Once you're done, msg me if you want me to merge for you, since there probably will be conflicts.

File: controls_pyqt/gui.py (hunk 4 above)
Comment (ayambabu23) [reply controls_pyqt/gui.py:229]: I don't see how it is a typo; I would like you to please explain this to me.

**Pull request merged**